lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,500 @@
+        private void UpdateState(string taskId, TaskStateEvent taskEvent)
+        {
+            if (!_tasks.ContainsKey(taskId))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] doesn't exist.", taskId);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+            GetTaskInfo(taskId).Item1.MoveNext(taskEvent);
+        }
+
+        /// <summary>
+        /// Checks if all the tasks are running.
+        /// </summary>
+        /// <returns></returns>
+        internal bool AreAllTasksAreRunning()
+        {
+            return _tasks.All(t => t.Value.Item1.CurrentState == StateMachine.TaskState.TaskRunning) &&
+                _runningTasks.Count == _totalExpectedTasks;
+        }
+
+        /// <summary>
+        /// Checks if all the tasks are completed.
+        /// </summary>
+        /// <returns></returns>
+        internal bool AreAllTasksAreCompleted()
+        {
+            return AllInTheState(StateMachine.TaskState.TaskCompleted) && _tasks.Count == _totalExpectedTasks && _runningTasks.Count == 0;
+        }
+
+        /// <summary>
+        /// This method is called when receiving either IFailedEvaluator or IFailedTask event
+        /// Driver tries to close all the running tasks and clean the running task collection in the end.
+        /// If all the tasks are running, the total number of running tasks should be _totalExpectedTasks -1
+        /// If this happens before all the tasks are running, then the total number of running tasks should smaller than _totalExpectedTasks -1
+        /// If this happens when no task is running, the total number of running tasks could be 0
+        /// </summary>
+        internal void CloseAllRunningTasks(string closeMessage)
+        {
+            Logger.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "Closing [{0}] running tasks.", _runningTasks.Count));
+            foreach (var runningTask in _runningTasks.Values)
+            {
+                runningTask.Dispose(Encoding.UTF8.GetBytes(closeMessage));
+                UpdateState(runningTask.Id, TaskStateEvent.WaitingTaskToClose);
+            }
+            _runningTasks.Clear();
+        }
+
+        /// <summary>
+        /// This method is called when receiving an IRunningTask event but system is either in ShuttingDown or Fail state.
+        /// In this case, the task should not be added in Running Tasks yet.
+        /// Change the task state to TaskRunning if it is still in TaskSubmitted state
+        /// Closes the IRunningTask 
+        /// Then move the task state to WaitingTaskToClose
+        /// Throw IMRUSystemException if runningTask is null or the running task is already added in the running task collection
+        /// </summary>
+        /// <param name="runningTask"></param>
+        /// <param name="closeMessage"></param>
+        internal void CloseRunningTask(IRunningTask runningTask, string closeMessage)
```

Comment:
I think we should coin a term for the state of all Tasks such that we can distinguish it in naming functions. This doesn't seem like a good name to me, but I'm not sure how I would rename it.